File: clients/go/zawgyi_unicode_markov_model.go
package myanmartools

import (
	"bufio"
	"fmt"
	"math"
)

const (
	//ModelBinaryTag Magic number used to identify this object in byte streams. (Reads in ASCII as "UZMODEL")
	ModelBinaryTag = 0x555A4D4F44454C20
	//StdCp0 Standard Myanmar code point range before digits
	StdCp0 int = '\u1000'
	//StdCp1 Standard Myanmar code point range before digits
	StdCp1 int = '\u103F'
	//AftCp0 standard myanmar code point range after digits
	AftCp0 int = '\u104A'
	//AftCp1 standard myanmar code point range after digits
	AftCp1 int = '\u109F'
	//ExaCp0 extended myanmar code point range A
	ExaCp0 int = '\uAA60'
	//ExaCp1 extended myanmar code point range A
	ExaCp1 int = '\uAA7F'
	//ExbCp0 extended Myanmar code point range B
	ExbCp0 int = '\uA9E0'
	//ExbCp1 extended Myanmar code point range B
	ExbCp1 int = '\uA9FF'
	//SpcCp0 unicode space character
	SpcCp0 int = '\u2000'
	//SpcCp1 unicode space character
	SpcCp1 int = '\u200B'

	//StdOffset index into Markov Nodes
	StdOffset int = 1
	//AftOffset index into Markov Nodes
	AftOffset int = StdOffset + StdCp1 - StdCp0 + 1
	//ExaOffset index into Markov Nodes
	ExaOffset int = AftOffset + AftCp1 - AftCp0 + 1
	//ExbOffset index into Markov Nodes
	ExbOffset int = ExaOffset + ExaCp1 - ExaCp0 + 1
	//SpcOffset index into Markov Nodes
	SpcOffset int = ExbOffset + ExbCp1 - ExbCp0 + 1
	//EndOffset index into Markov Nodes
	EndOffset int = SpcOffset + SpcCp1 - SpcCp0 + 1

	//Ssv originally stands for state set version

	//SsvStdExaExbSpc include Myanmar, Extended A, Extended B and space-like
	SsvStdExaExbSpc int = 0
	//SsvStdExaExb same as above but no space-like code points
	SsvStdExaExb int = 1
	//SsvCount ssv count
	SsvCount int = 2
)

//ZawgyiUnicodeMarkovModel interface
type ZawgyiUnicodeMarkovModel interface {
	GetIndexForCodePoint(cp int, ssv int) int
	GetSize(ssv int) int
	Predict(input string) float64
}
type zawgyiUnicodeMarkovModel struct {
	classifier BinaryMarkov
	ssv        int
}

func (*zawgyiUnicodeMarkovModel) GetIndexForCodePoint(cp int, ssv int) int {
	// fmt.Printf("Getting Index for code point %d\n", cp)
	if StdCp0 <= cp && cp <= StdCp1 {
		// fmt.Printf("Code point %d between StdCp0 and StdCp1\n", cp)
		return cp - StdCp0 + StdOffset
	}

	if AftCp0 <= cp && cp <= AftCp1 {
		// fmt.Printf("Code point %d between AftCp0 and AftCp1\n", cp)
		return cp - AftCp0 + AftOffset
	}

	if ExaCp0 <= cp && cp <= ExaCp1 {
		// fmt.Printf("Code point %d between ExaCp0 and ExaCp1\n", cp)
		return cp - ExaCp0 + ExaOffset
	}

	if ExbCp0 <= cp && cp <= ExbCp1 {
		// fmt.Printf("Code point %d between ExbCp0 and ExbCp1\n", cp)
		return cp - ExbCp0 + ExbOffset
	}

	if ssv == SsvStdExaExbSpc && SpcCp0 <= cp && cp <= SpcCp1 {
		// fmt.Printf("Code point %d between SpcCp0 and SpcCp1 and code point == SsvStdExaExbSpc\n", cp)
		return cp - SpcCp0 + SpcOffset
	}
	// fmt.Printf("Code point is %d\n", cp)
	return 0
}

func (*zawgyiUnicodeMarkovModel) GetSize(ssv int) int {
	if ssv == SsvStdExaExbSpc {
		return EndOffset
	}
	return SpcOffset
}

func (zumm *zawgyiUnicodeMarkovModel) Predict(input string) float64 {

	// Start at the base state
	var prevState int = 0

	var totalDelta float64 = 0.0
	seenTransition := false

	for offset := 0; offset <= len([]rune(input)); offset++ {
		var currState int
		var cp int
		if offset == len([]rune(input)) {
			currState = 0
			cp = 0
		} else {
			cp = int([]rune(input)[offset])
			// fmt.Printf("Char code %d\n", cp)
			currState = zumm.GetIndexForCodePoint(cp, zumm.ssv)
			// fmt.Printf("Index for code point %d\n", currState)
		}

		//Ignore 0-to-0 transitions
		if prevState != 0 || currState != 0 {
			delta := zumm.classifier.GetLogProbabilityDifference(prevState, currState)
			// fmt.Printf("Log Probability Difference: %.20f\n", delta)
			totalDelta += float64(delta)
			seenTransition = true
		}
		prevState = currState
	}

	// Special case: if there is no signal, return -Infinity,
	// which will get interpreted by users as strong Unicode.
	// This happens when the input string contains no Myanmar-range code points.
	if !seenTransition {
		return math.Inf(-1)
	}

	// result = Pz/(Pu+Pz)
	//        = exp(logPz)/(exp(logPu)+exp(logPz))
	//        = 1/(1+exp(logPu-logPz))

	return 1.0 / (1.0 + math.Exp(totalDelta))
}

//NewZawgyiUnicodeMarkovModel creates and
//returns ZawgyiUnicodeMarkovModel implementation for classifier and ssv
func NewZawgyiUnicodeMarkovModel(classifier BinaryMarkov, ssv int) ZawgyiUnicodeMarkovModel {
	return &zawgyiUnicodeMarkovModel{classifier: classifier, ssv: ssv}
}

//NewZawgyiUnicodeMarkovModelFromReader creates and
//returns ZawgyiUnicodeMarkovModel from specified dat file
func NewZawgyiUnicodeMarkovModelFromReader(datReader *bufio.Reader) ZawgyiUnicodeMarkovModel {

	modelBinaryTag := readLong(datReader)
	var ssv int
	var classifier BinaryMarkov
	if ModelBinaryTag != modelBinaryTag {
		panic(fmt.Errorf("Unexpected magic number: expected %X but got %X", ModelBinaryTag, modelBinaryTag))
	}
	binaryVersion := readInt(datReader)
	if binaryVersion == 1 {
		ssv = SsvStdExaExbSpc
	} else if binaryVersion == 2 {
		ssv = readInt(datReader)
	} else {
		panic(fmt.Errorf("Unexpected serial version number; expected 1 or 2 but got %X", binaryVersion))
	}

	if ssv < 0 || ssv >= SsvCount {
		panic(fmt.Errorf("Unexpected value in ssv position; expected 0 or 1 but %X", ssv))
	}
	classifier = NewBinaryMarkov(datReader)
	return &zawgyiUnicodeMarkovModel{classifier: classifier, ssv: ssv}
}

File: clients/php/src/ZawgyiUnicodeMarkovModel.php
<?php
/**
 * Created by SteveNay.
 * User: NayLinAung
 * Date: 5/24/2018
 * Time: 3:39 PM
 */

namespace Googlei18n\MyanmarTools;

use Exception;

/**
 * A Markov model to predict whether a string is more likely Unicode or Zawgyi.
 * <p>
 * <p>Internally, this class maintains two Markov chains, one representing Unicode and the other
 * representing Zawgyi. An input string is evaluated against both chains, and the chain that returns
 * the higher probability becomes the prediction.
 * <p>
 * <p>A string is evaluated as a sequence of transitions between states, including transitions to
 * the edges of the string. For example, the string "ABC" contains 4 transitions: NULL to A, A to B,
 * B to C, and C to NULL.
 * <p>
 * <p>For the purposes of Unicode/Zawgyi detection, all characters are treated as the NULL state
 * except for characters in the Myanmar script or characters in the Unicode whitespace range U+2000
 * through U+200B.
 * <p>
 * <p>This class is tested in the google-myanmar-tools-data package because this implementation does
 * not include any logic for building a new instance other than from a binary data stream.
 */
class ZawgyiUnicodeMarkovModel
{
    /** Magic number used to identify this object in byte streams. (Reads in ASCII as "BMARKOV ") */
    const BINARY_TAG = "555a4d4f44454c20";

    // Standard Myanmar code point range before digits
    const STD_CP0 = 0x1000;
    const STD_CP1 = 0x103F;

    // Standard Myanmar code point range after digits
    const AFT_CP0 = 0x104A;
    const AFT_CP1 = 0x109F;

    // Extended Myanmar code point range A
    const EXA_CP0 = 0xAA60;
    const EXA_CP1 = 0xAA7F;

    // Extended Myanmar code point range B
    const EXB_CP0 = 0xA9E0;
    const EXB_CP1 = 0xA9FF;

    // Unicode space characters
    const SPC_CP0 = 0x2000;
    const SPC_CP1 = 0x200B;

    // Indices into Markov nodes
    const STD_OFFSET = 1;
    const AFT_OFFSET = self::STD_OFFSET + self::STD_CP1 - self::STD_CP0 + 1;
    const EXA_OFFSET = self::AFT_OFFSET + self::AFT_CP1 - self::AFT_CP0 + 1;
    const EXB_OFFSET = self::EXA_OFFSET + self::EXA_CP1 - self::EXA_CP0 + 1;
    const SPC_OFFSET = self::EXB_OFFSET + self::EXB_CP1 - self::EXB_CP0 + 1;
    const NUM_STATES = self::SPC_OFFSET + self::SPC_CP1 - self::SPC_CP0 + 1;

    // BinaryMarkov
    private $classifier;

    // TODO: Not currently used
    private $ssv;

    /**
     * Creates an instance from a binary data stream.
     *
     * @throws Exception If there was a problem reading the data.
     */
    public function __construct($stream)
    {
        // Check magic number and serial version number
        $binaryTag = unpack("H*", fread($stream, 8))[1];

        if ($binaryTag !== self::BINARY_TAG) {
            throw new Exception(
                sprintf(
                    "Unexpected magic number; expected %016X but got %016X", self::BINARY_TAG, $binaryTag));
        }

        $binaryVersion = (int) unpack("H*", fread($stream, 4))[1];

        if ($binaryVersion === 1) {
            $this->ssv = 0;
        } else if ($binaryVersion == 2) {
            // TODO: Support nonzero SSV if needed in the future
            $this->ssv = (int) unpack("H*", fread($stream, 4))[1];
            if ($this->ssv != 0) {
                throw new Exception(sprintf("Unsupported ssv: %d", $this->ssv));
            }
        } else {
            throw new Exception(
                sprintf(
                    "Unexpected serial version number; expected 1 or 2 but got %08X",
                    $binaryVersion));
        }

        // check for utf-8 encoding.
        if (mb_internal_encoding() != "UTF-8") {
            $message = sprintf("Encoding must be UTF-8; got %s. Please call mb_internal_encoding('UTF-8')", mb_internal_encoding());
            throw new Exception($message);
        }

        $this->classifier = new BinaryMarkov($stream);
    }

    /**
     * Returns the index of the state in the Markov chain corresponding to the given code point.
     * <p>
     * <p>Code points in the standard Myanmar range, Myanmar Extended A, Myanmar Extended B, and
     * Unicode Whitespace each have a unique state assigned to them. All other code points are mapped
     * to state 0.
     * <p>
     * <p>Package-private so that the builder can use this method.
     *
     * @param int $cp The code point to convert to a state index.
     * @return The index of the state in the Markov chain. 0 <= state < getSize()
     */
    public static function getIndexForCodePoint(int $cp)
    {
        if (self::STD_CP0 <= $cp && $cp <= self::STD_CP1) {
            return $cp - self::STD_CP0 + self::STD_OFFSET;
        }

        if (self::AFT_CP0 <= $cp && $cp <= self::AFT_CP1) {
            return $cp - self::AFT_CP0 + self::AFT_OFFSET;
        }

        if (self::EXA_CP0 <= $cp && $cp <= self::EXA_CP1) {
            return $cp - self::EXA_CP0 + self::EXA_OFFSET;
        }

        if (self::EXB_CP0 <= $cp && $cp <= self::EXB_CP1) {
            return $cp - self::EXB_CP0 + self::EXB_OFFSET;
        }

        if (self::SPC_CP0 <= $cp && $cp <= self::SPC_CP1) {
            return $cp - self::SPC_CP0 + self::SPC_OFFSET;
        }

        return 0;
    }

    /**
     * The number of states in the Markov chain.
     */
    public static function getSize()
    {
        return self::NUM_STATES;
    }

    /**
     * @param string $input
     * @param bool $verbose
     * @return float
     */
    public function predict(string $input, bool $verbose = false): float
    {
        if ($verbose) {
            printf("Running detector on string: %s\n", $input);
        }

        // Start at the base state
        $prevCp    = 0;
        $prevState = 0;

        $totalDelta     = 0.0;
        $seenTransition = false;

        $inputBreakIterator = \IntlBreakIterator::createCodePointInstance();
        $inputBreakIterator->setText($input);

        do {
            $idx = $inputBreakIterator->next();
            $cp = ($idx === \IntlBreakIterator::DONE) ? 0 : $inputBreakIterator->getLastCodePoint();
            $currState = self::getIndexForCodePoint($cp);
            // Ignore 0-to-0 transitions
            if ($prevState != 0 || $currState != 0) {
                $delta = $this->classifier->getLogProbabilityDifference($prevState, $currState);
                if ($verbose) {
                    printf("U+%04X -> U+%04X: delta=%8.4f ", $prevCp, $cp, $delta);
                    for ($i = 1; $i < abs($delta); $i++) {
                        printf("!");
                    }
                    print("\n");
                }
                $totalDelta     += $delta;
                $seenTransition = true;
            }

            $prevCp    = $cp;
            $prevState = $currState;
        } while ($cp !== 0);

        if ($verbose) {
            printf("Final: delta=%.4f\n", $totalDelta);
        }

        // Special case: if there is no signal, return -Infinity,
        // which will get interpreted by users as strong Unicode.
        // This happens when the input string contains no Myanmar-range code points.
        if (!$seenTransition)
            return INF;

        // result = Pz/(Pu+Pz)
        //        = exp(logPz)/(exp(logPu)+exp(logPz))
        //        = 1/(1+exp(logPu-logPz))
        return 1.0 / (1.0 + exp($totalDelta));
    }
}